Checkout: Keep VAT country in sync with parent country

sirbrillig opened the pull request:
#### Background

When VAT form support was added to checkout in https://github.com/Automattic/wp-calypso/pull/71285, there was a subtle bug. Checkout only has one country code form field and it has three uses: the shopping cart endpoint tax country, the cached domain contact endpoint country, and (now) the VAT data endpoint country. Normally, all three of these values are the same, but the VAT form has to have its own local version of the data because there are cases (Northern Ireland) where the country code for the VAT form will differ from the other two (although the shopping cart tax data will also use this value, it is still stored in a separate place in the code).

Here's the bug: if the cached contact details endpoint returns one country code, but the VAT data endpoint returns a different country code, we need to make sure that we use the correct one. In this case that's the cached contact details endpoint code because that is the one that the checkout form uses; if we used the VAT country code, which is essentially a hidden field, there would be no way to see or change it and you'd get strange validation errors from the VAT endpoint.

#### Proposed Changes

To prevent this bug, in this PR we make a more concerted effort to keep the "hidden" VAT country code field (it's not actually a field; just a variable) in sync with the `VatForm`'s parent country code field. Any time the parent's country code field (the only one on the form) changes, we update the VAT country code to match. We only break that sync if the Northern Ireland checkbox is checked.

In addition, this fixes a bug where changing the country from a VAT-supported country to a non-VAT-supported country would still send any existing VAT data to the shopping cart. This was happening because while the VAT data is cleared when the "Add VAT details" checkbox is unchecked, that side effect was not happening when the `VatForm` itself was unmounted as a result of the country no longer being applicable. Here we fix that issue by always rendering `VatForm` and having it hide itself when the country does not apply so it can continue to run side effects as needed.

#### Testing Instructions

This PR adds tests to reproduce the bugs mentioned above and prove that they are fixed. This also fixes several other tests that weren't quite testing what they claimed to be testing.

You can test this manually by using checkout to set a cached country code in the EU (like "Spain") and clicking "Continue"; then using the `/me/purchases/vat-details` page to set a VAT country that's different (like `GB`, along with valid VAT details; you can use the Vat ID `553557881` if the store is sandboxed).

After that's set up, visit checkout and you'll see that the country field shows the cached country (eg: "Spain"), but if you try to submit the form and then look at the contact step summary. Before this PR you'll see it reads `GB`; after this PR you'll see `ES`.

Finally, change the country code to "United States" and set a valid postal code like `90210`. Press "Continue" and verify that the summary shows `US`.

sirbrillig committed:
Keep VAT country code in sync with parent form country code

sirbrillig committed:
Add tests for VAT address

sirbrillig committed:
Make sure to init country in VatForm when box is checked

sirbrillig committed:
Fix bugs in pre-filled VAT form tests

sirbrillig committed:
Add test for different pre-filled contact and VAT country

sirbrillig committed:
Wait for the cart to load in tests by looking for continue button

sirbrillig committed:
Fix type errors in checkout test utils

sirbrillig committed:
Remove address code in test

sirbrillig committed:
Fix price_before_discounts_integer in test utils

sirbrillig committed:
Allow GET vat-info requests to be always made in contact tests

sirbrillig committed:
Fix some more type errors in checkout test utils

sirbrillig committed:
Add tests for shopping-cart endpoint VAT data

pottedmeat commented:
- I cleared the VAT details for my testing account.
- I switched my billing information to Spain on the checkout form.
- I went to the VAT details edit page.
- I set country to GB and filled out VAT details.
- I went back to checkout and reloaded it
- "Add VAT details" is checked
- I see all my VAT details and the country is set to Spain
- I switched it to United Kingdom and gave it a UK postal code
- It saved correctly
- Reloading the page shows the same VAT details

pottedmeat reviewed (approved):
Code looks good.

My only feedback is that `setVatDetailsInForm` is updated sometimes with `savedVatDetails` and sometimes with `vatDetailsInForm`. I feel like I understand why, but it would be nice to have additional comments explaining it.

sirbrillig committed:
Always render VAT form if config enables it

sirbrillig committed:
Refactor VatForm to hide itself if country unsupported

This allows the form to trigger side effects (removing VAT details from
the store) if the country becomes unsupported.

sirbrillig commented:
Thanks! I refactored the way that `VatForm` pre-fills its fields to fix the issue you described in https://github.com/Automattic/wp-calypso/pull/72734#issuecomment-1407433999 and also added a test for that bug. However, I made the fixes in https://github.com/Automattic/wp-calypso/pull/72734 and tried to backport them here and they don't work here. I don't have time to figure out why just now so please ignore this PR for the time being and just test that one.

sirbrillig committed:
Fix typo from rebase

sirbrillig committed:
Create helper functions to mock VAT endpoints for tests

sirbrillig commented:
Ok, I think I've got this working correctly again.

pottedmeat reviewed (approved):
Reran my tests. Differences from first run:

- Billing information was automatically expanded after filling out the form on the VAT details edit page but before changing it from Spain to UK

Everything else passed as before.
